### ISSUE TITLE
False positive with S1854

### ISSUE THREAD

**Issue by cdanek** [SonarSource/sonar-dotnet#6154]
Product: SolarLint 6.8.0.53188
IDE: VS 2022 CE 17.0.4 Preview 2
Language: C#
Rule: S1854 - "Remove this useless assignment to local variable"

I am using the prefix increment operator in code that sends an integer to a method. Here's a MRE:

```
       public void MyMethod()
        {
            int i = 0;
            MyMethod2(++i);
        }
        public void MyMethod2(int i)
        {
            _ = i;
        }
```

If there are hundreds of calls to `MyMethod2(int i)`, making the last and only last "`i+1`" is a code smell in itself:

![image](https://user-images.githubusercontent.com/21341843/190937664-8f25bd15-2e1e-4df6-ab43-b9a2d70a2d2c.png)

(this would be difficult to spot visually, and adding another line to the end with the prefix operator would introduce a bug).

**Comment by mary-georgiou-sonarsource** [comment]
Thank you for opening this issue, @cdanek. Your feedback helps us improve our products.

We will look at it as soon as possible. After we confirm the issue, we will add labels and put it in our backlog. Once it's triaged by us, you can open a PR for it.

**Comment by mary-georgiou-sonarsource** [comment]
Hello @cdanek!
Thanks a lot for raising an issue.

Could you please provide a code example and mark where is precisely the False Positive issue raised?

> If there are hundreds of calls to MyMethod2(int i), making the last and only last "i+1" is a code smell in itself.

Why the last call would be a code smell?

**Comment by cdanek** [comment]
> Hello @cdanek! Thanks a lot for raising an issue.
> 
> Could you please provide a code example and mark where is precisely the False Positive issue raised?
> 
> > If there are hundreds of calls to MyMethod2(int i), making the last and only last "i+1" is a code smell in itself.
> 
> Why the last call would be a code smell?

There is a code sample above. To make it a bit more obvious:

```
       public void MyMethod()
        {
            int i = 0;
            MyMethod2(++i);
            MyMethod2(++i);
            MyMethod2(++i);
            MyMethod2(++i);
            MyMethod2(++i); // S1854
        }
        public void MyMethod2(int i)
        {
            _ = i;
        }
```

The last call is a code smell because it is difficult to see visually (look close, it's not "plus plus i" it's "i plus one", and calls which depend on correct ordering are likely to introduce bugs. Imagine if "MyMethod2" above were something like "add to order", where the last (and only) the last call needed to be different from the rest. Now imagine you come back to this codebase later, visually see how the call works, copy and paste a few more lines to the bottom:

```
       public void MyMethod()
        {
            int i = 0;
            MyMethod2(++i);
            MyMethod2(++i);
            MyMethod2(++i);
            MyMethod2(i+1);
            MyMethod2(++i); // Oops, bug here.
            MyMethod2(++i);
            MyMethod2(++i);
            MyMethod2(++i);
        }
        public void MyMethod2(int i)
        {
            _ = i;
        }
```

**Comment by cdanek** [comment]
Hi @mary-georgiou-sonarsource, thank you for the comments and responses. :)

I suppose I understand how the rule is crafted and attacking the problem, and you're right, that the last `++i` wouldn't have any further use in the calling method. I suppose I see it differently, though, from a cleanliness (or OCD ha ha) perspective: It would "itch" to have all the calls use `++i` except for the last which used `i+1`. 

No worries, and thanks for the investigation. Closing the issue, have a good day!
